### DISCUSSION TITLE
Running Emanote in Raspberry

### DISCUSSION THREAD

**Discussion by bolt12** [srid/emanote#239] [Q&A] [answered]
I would like to host emanote on my personal NixOS RPI but unfortunately I couldn't use either docker or nix installation since there are architecture differences. I wonder how much trouble would be to make the project support such architectures.

Thank you for everyone's time!

**Comment by srid** [comment]
@bolt12 Try with https://github.com/srid/emanote/commit/9dc23037ddbba7955d6865b8967a7d1ba21c5032

You can also talk to @573 who has done this in the past: https://github.com/srid/neuron/discussions/66

**Reply by srid** [reply]
@bolt12 You would have better luck asking in NixOS/Nix discussion channels though.

**Reply by bolt12** [reply]
I see! Thank you for the pointers!

**Reply by bolt12** [reply]
I tried with https://github.com/srid/emanote/commit/9dc23037ddbba7955d6865b8967a7d1ba21c5032 and made some progress. I think there's the similar missing for the haskell-tailwind package:

```
unpacking 'https://api.github.com/repos/srid/tailwind-haskell/tarball/51c69972f3bdd28727f4a0473625cd05c8f7c4f3'...
error: attribute 'aarch64-linux' missing, at /nix/store/f87i396f7z2ffp37b1l8fqf3fmqj9wxq-source/flake.nix:92:21
(use '--show-trace' to show detailed location information)
building Nix...
error: attribute 'aarch64-linux' missing, at /nix/store/f87i396f7z2ffp37b1l8fqf3fmqj9wxq-source/flake.nix:92:21
(use '--show-trace' to show detailed location information)
building the system configuration...
error: attribute 'aarch64-linux' missing, at /nix/store/f87i396f7z2ffp37b1l8fqf3fmqj9wxq-source/flake.nix:92:21
(use '--show-trace' to show detailed location information)
```

**Reply by srid** [reply]
You can try f942b3a6c37dbf6bdcc73a85fdf33f58eb7b9129

**Reply by bolt12** [reply]
Thank you for your help trying getting this to work! We are making progress, but apparently now:

```
unpacking 'https://api.github.com/repos/srid/tailwind-nix/tarball/e27e11a66ae26533b265eb826eec357fbf481c92'...
error: attribute 'aarch64-linux' missing, at /nix/store/l3a3jgm9xxq022366icgbfywxcswn4l8-source/flake.nix:58:19
(use '--show-trace' to show detailed location information)
```

It seems a similar change needs to be made in srid/tailwind-nix .

**Reply by srid** [reply]
Okay e1519a56222dcc66804a76f19fccc3a75d83af9e should help.

Btw, you can also fork these repos and override the flake inputs to get it to build (then open PRs at the end).

**Reply by bolt12** [reply]
Yes, I will do that, sorry for making you do it instead, hopefully is just enabling this in the flake files :) 

Thank you once again for your help!

**Reply by bolt12** [reply]
I made progress but ultimately I've hit:

```
copying path '/nix/store/ky2gvkg7zrrifzrnj07c2pp0p9wyzlhy-cabal2nix-2.18.0' from 'https://cache.nixos.org'...
error: a 'aarch64-darwin' with features {} is required to build '/nix/store/7ddsmmw6prqrhlw39qm8618ga3rp8fp9-cabal2nix-emanote.drv', but I am a 'aarch64-linux' with features {benchmark, big-parallel, gccarch-armv8-a, kvm, nixos-test}
(use '--show-trace' to show detailed location information)
```

And I do not really know what I can do to solve this 😛

**Reply by srid** [reply]
@bolt12 I recommend asking in https://matrix.to/#/#community:nixos.org ... I have a feeling this can be solved.

**Reply by bolt12** [reply]
I asked here: https://matrix.to/#/!RjBlCIbsLDzHBIzmaA:nixos.org/$zJPbuC6r5ZYhx4PP_vtKYqapuW2pzEwaxSA8xoqNZ6g?via=nixos.org&via=matrix.org&via=nixos.dev

This is what I ended up with:

```nix


# List packages installed in system profile. To search, run:
# $ nix search wget
environment.systemPackages = let
   emanote = (import (pkgs.fetchFromGitHub {owner = "srid"; repo = "emanote"; rev = "master"; sha256 = "19xy042f7icmk41s206qbfi3fy4pw8sj5l6xgn6afwk2pzkhrx06"; })).defaultPackage.aarch64-linux;
 in
 with pkgs; [
...
emanote
];
```

This is a rather big package and I haven't run into any errors yet on my installation but hopefully I will have an answer soon :)

**Comment by bolt12** [comment, marked as answer]
Sorry for the delay on this, but I can say that I was able to make it work on my RPI. The derivation looks something like this:

```
environment.systemPackages = let
emanote = (import (pkgs.fetchFromGitHub {owner = "srid"; repo = "emanote"; rev = "master"; sha256 = "055i83av5lqiya8qlasw0laznz2mqag8arpg1d55bzc2bgz08hin"; })).defaultPackage.aarch64-linux;
in
with pkgs; [
...

emanote
];
```

My RPI only has 1 GB of RAM so I had to up the swap space to about 8GB and leave it building for like 2 days but it succeeded! Thank you for all the support @srid . Hope this thread helps other people out!

**Reply by srid** [reply]
@bolt12 Nice! I wonder if there is a way to provide cache for this (built in CI somewhere).

**Reply by srid** [reply]
@bolt12 How do you use Emanote in Raspberry Pi? Curious about your workflow ...

**Reply by bolt12** [reply]
> @bolt12 Nice! I wonder if there is a way to provide cache for this (built in CI somewhere).

One can try and setup something with https://www.cachix.org/ !

> @bolt12 How do you use Emanote in Raspberry Pi? Curious about your workflow ...

Good question, I am looking just for something were my notes can look pretty and allows me to navigate better or maybe even shape/improve the way I do take notes. I haven't really used emanote yet but looked at it and sounded the sort of thing I wanted :D
